src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -691,8 +691,15 @@ internal MethodSymbol TryFindDisposePatternMethod(BoundExpression expr, SyntaxNo
                                                     diagnostics,
                                                     out var disposeMethod);
 
-            if ((!hasAwait && disposeMethod?.ReturnsVoid == false)
-                || (hasAwait && disposeMethod?.ReturnType.TypeSymbol.IsNonGenericTaskType(Compilation) == false)
+            if (disposeMethod?.IsExtensionMethod == true)
+            {
+                // Extension methods should just be ignored, rather than rejected after-the-fact
+                // Tracked by https://github.com/dotnet/roslyn/issues/32767
+
+                // extension methods do not contribute to pattern-based disposal
+                disposeMethod = null;
```

Comment:
The `disposeMethod` is discarded here but what about any diagnostics that resulted from calling `PerformPatternMethodLookup`? Won't those still be in `diagnostics` and hence passed back up to the caller?

Reply:
In the `using` scenarios, we keep those diagnostics and add another error (didn't find a proper way of disposing). See `UsingPatternScopedExtensionMethodTest`, which produces some unnecessary diagnostics. I'm thinking to leave that as-is for now, with tracking issue.

In the `foreach` scenarios, the caller already uses a separate diagnostic bag. I'll add tests to demonstrate.